packages/commonwealth/server/util/roles.ts

```diff
@@ -0,0 +1,262 @@
+  if (permissions) {
+    roleFindOptions = {
+      where: {
+        [Op.and]: [
+          { chain_id },
+          {
+            name: {
+              [Op.or]: permissions,
+            },
+          },
+        ],
+      },
+      include: [
+        {
+          model: models.RoleAssignment,
+          ...findOptions,
+        },
+      ],
+      order: [
+        sequelize.literal(
+          `CASE WHEN name='admin' THEN 0 WHEN name='moderator' THEN 1 ELSE 2 END`
+        ),
+      ],
+    };
+  } else {
+    roleFindOptions = {
+      where: {
+        chain_id,
+      },
+      include: [
+        {
+          model: models.RoleAssignment,
+          ...findOptions,
+        },
+      ],
+      order: [
+        sequelize.literal(
+          `CASE WHEN name='admin' THEN 0 WHEN name='moderator' THEN 1 ELSE 2 END`
+        ),
+      ],
+    };
+  }
+
+  const communityRole = await models.CommunityRole.findOne(roleFindOptions);
+  let role: RoleInstanceWithPermission;
+  if (communityRole) {
+    // Retrieve the highest role as it will be the highest permission role for the address_id
+    if (communityRole.RoleAssignments.length > 0) {
+      const roleAssignment = communityRole.RoleAssignments[0];
+      role = new RoleInstanceWithPermission(
+        roleAssignment.toJSON(),
+        chain_id,
+        communityRole.name,
+        communityRole.allow,
+        communityRole.deny
+      );
+      return role;
+    } else {
+      return null;
+    }
+  } else {
+    return null;
+  }
+}
+
+export async function createDefaultCommunityRoles(
+  models: DB,
+  chain_id: string
+): Promise<void> {
+  // Create default roles
+  try {
+    await models.CommunityRole.create({
+      chain_id,
+      name: 'member',
+      allow: BigInt(0),
+      deny: BigInt(0),
+    });
+    await models.CommunityRole.create({
+      chain_id,
+      name: 'moderator',
+      allow: BigInt(0),
+      deny: BigInt(0),
+    });
+    await models.CommunityRole.create({
+      chain_id,
+      name: 'admin',
+      allow: BigInt(0),
+      deny: BigInt(0),
+    });
+  } catch (error) {
+    throw new Error(`Couldn't create default community roles ${error}`);
+  }
+}
+
+export async function createRole(
+  models: DB,
+  address_id: number,
+  chain_id: string,
+  role_name?: Permission,
+  is_user_default?: boolean,
+  transaction?: Transaction
+): Promise<RoleInstanceWithPermission> {
+  if (role_name === undefined) {
+    role_name = 'member';
+  }
+
+  // check if role is already assigned to address
+  const role = await findOneRole(models, { where: { address_id } }, chain_id, [
+    role_name,
+  ]);
+  if (role) {
+    // if role is already assigned to address, return highest role this address has on that chain
+    return findOneRole(models, { where: { address_id } }, chain_id);
+  }
+
+  // Get the community role that has given chain_id and name if exists
+  let community_role;
+  community_role = await models.CommunityRole.findOne({
```

Comment:
Can use `findOrCreate` here -- however, it should always exist, because communities get `admin`, `moderator`, and `member` by default, and no others are allowed.

Reply:
this makes sense